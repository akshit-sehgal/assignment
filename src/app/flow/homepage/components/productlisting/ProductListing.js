import React from 'react';
import PropTypes from 'prop-types';
import ProductListItem from './ProductListItem';
import './style.scss';
import Shimmer from '../../../../shared/components/shimmer/Shimmer';
import { LISTING_LABELS } from './constants';
import { ROUTE_NAMES } from '../../../../RouteConstants';

class ProductListing extends React.PureComponent {
    navigateToDetailsPage = (e) => {
        const { navigate } = this.props;
        const id = e.currentTarget.getAttribute('data-id');
        navigate(`${ROUTE_NAMES.PRODUCT_DETAILS_PAGE}/${id}`);
    }
    generateProductsList = () => {
        const { productsList } = this.props;
        if (!productsList.length) return;
        return productsList.map((item) => (
            <ProductListItem
                key={item.id}
                navigate={this.navigateToDetailsPage}
                {...item}
            />
        ));
    }
    generateLoadMore = () => {
        const { nextPage, fetchProducts } = this.props;
        return (
            nextPage ?
                <button className="products__more-btn" onClick={fetchProducts}>{LISTING_LABELS.MORE_BTN}</button> :
                <div className="product__no-more">{LISTING_LABELS.NOTHING_MORE}</div>
        );
    }
    componentDidMount() {
        const { productsList, fetchProducts } = this.props;
        if (!productsList.length) {
            fetchProducts();
        }
        window.scrollTo(0, 0);
    }

    render() {
        const { isFetchingProducts } = this.props;
        return (
            <div className="products-container section-container">
                <div className="products__heading">{LISTING_LABELS.HEADING}</div>
                <hr />
                <div className="products__list">
                    {this.generateProductsList()}
                </div>
                {isFetchingProducts && <Shimmer count={10} />}
                {this.generateLoadMore()}
            </div>
        );
    }

}
ProductListing.propTypes = {
    productsList: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        img: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        price: PropTypes.string.isRequired
    })).isRequired,
    isFetchingProducts: PropTypes.bool.isRequired,
    nextPage: PropTypes.number.isRequired,
    fetchProducts: PropTypes.func.isRequired
}
export default ProductListing;