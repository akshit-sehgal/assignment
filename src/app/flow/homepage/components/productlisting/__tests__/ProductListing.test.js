import React from 'react';
import { shallow } from 'enzyme';
import ProductListing from '../ProductListing';
import { mockProductListingData } from '../__mocks__/productListing.mock';

describe('Testing <ProductListing />', () => {
    let props;
    window.scrollTo = jest.fn();
    beforeEach(() => {
        props = {
            productsList: mockProductListingData.productsList,
            nextPage: mockProductListingData.nextPage,
            isFetchingProducts: false,
            fetchProducts: jest.fn(),
            navigate: jest.fn()
        }
    });
    it('should render the default view', () => {
        const wrapper = shallow(<ProductListing {...props} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ProductListItem').length).toEqual(props.productsList.length);
        expect(wrapper.find('.products__more-btn').length).toEqual(1);
        wrapper.setProps({
            productsList: [],
            nextPage: 0,
            isFetchingProducts: true
        });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ProductListItem').length).toEqual(0);
        expect(wrapper.find('.product__no-more').length).toEqual(1);
        expect(wrapper.find('Shimmer').length).toEqual(1);
    });
    it('should fetchProducts on mount', () => {
        props.productsList = [];
        const wrapper = shallow(<ProductListing {...props} />);
        expect(props.fetchProducts).toHaveBeenCalledTimes(1);
        expect(window.scrollTo).toHaveBeenCalledWith(0, 0);
    });
    it('should trigger navigation', () => {
        const wrapper = shallow(<ProductListing {...props} />);
        const e = {
            currentTarget: {
                getAttribute: jest.fn((val) => '1')
            }
        };
        wrapper.find('ProductListItem').at(0).prop('navigate')(e);
        expect(props.navigate).toHaveBeenCalledWith(`/pdp/1`)
    });

});