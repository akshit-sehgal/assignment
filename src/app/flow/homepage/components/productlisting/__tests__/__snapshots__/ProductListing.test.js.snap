// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <ProductListing /> should render the default view 1`] = `
<div
  className="products-container section-container"
>
  <div
    className="products__heading"
  >
    All Products
  </div>
  <hr />
  <div
    className="products__list"
  >
    <ProductListItem
      id={41}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/6/12/2570456/2570456_90794ecf-e3f4-4896-b466-efe108372d5a_720_720.jpg.webp"
      key="41"
      name="Product 41"
      navigate={[Function]}
      price="1000"
    />
    <ProductListItem
      id={42}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/11/6/5723707/5723707_6ffc1993-cbc3-45d1-8b24-3619c9d39803_1280_1280.webp"
      key="42"
      name="Product 42"
      navigate={[Function]}
      price="1100"
    />
    <ProductListItem
      id={43}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/10/22/12590409/12590409_0285b3d3-ab8f-41ca-806c-ff2290ca0f96_700_700.webp"
      key="43"
      name="Product 43"
      navigate={[Function]}
      price="1200"
    />
    <ProductListItem
      id={44}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/10/23/290238/290238_7e65adb0-c425-4558-bf2f-ea876cc43c5d_640_640.jpg.webp"
      key="44"
      name="Product 44"
      navigate={[Function]}
      price="1300"
    />
    <ProductListItem
      id={45}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/11/27/6041270/6041270_46edf22c-7ac5-4b01-b166-ca9dbd3444be_869_869.jpg.webp"
      key="45"
      name="Product 45"
      navigate={[Function]}
      price="1400"
    />
    <ProductListItem
      id={46}
      img="https://ecs7.tokopedia.net/img/cache/200-square/attachment/2019/11/6/157302369912918/157302369912918_6fbb518f-f0fc-4241-96d3-1dca699f4c15.png.webp"
      key="46"
      name="Product 46"
      navigate={[Function]}
      price="1500"
    />
    <ProductListItem
      id={47}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/6/12/2570456/2570456_90794ecf-e3f4-4896-b466-efe108372d5a_720_720.jpg.webp"
      key="47"
      name="Product 47"
      navigate={[Function]}
      price="1600"
    />
    <ProductListItem
      id={48}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/11/6/5723707/5723707_6ffc1993-cbc3-45d1-8b24-3619c9d39803_1280_1280.webp"
      key="48"
      name="Product 48"
      navigate={[Function]}
      price="1700"
    />
    <ProductListItem
      id={49}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/10/22/12590409/12590409_0285b3d3-ab8f-41ca-806c-ff2290ca0f96_700_700.webp"
      key="49"
      name="Product 49"
      navigate={[Function]}
      price="1800"
    />
    <ProductListItem
      id={50}
      img="https://ecs7.tokopedia.net/img/cache/200-square/product-1/2019/10/23/290238/290238_7e65adb0-c425-4558-bf2f-ea876cc43c5d_640_640.jpg.webp"
      key="50"
      name="Product 50"
      navigate={[Function]}
      price="1900"
    />
  </div>
  <button
    className="products__more-btn"
    onClick={[MockFunction]}
  >
    Load More
  </button>
</div>
`;

exports[`Testing <ProductListing /> should render the default view 2`] = `
<div
  className="products-container section-container"
>
  <div
    className="products__heading"
  >
    All Products
  </div>
  <hr />
  <div
    className="products__list"
  />
  <Shimmer
    count={10}
    showImageHolder={true}
    showLabelsHolder={true}
  />
  <div
    className="product__no-more"
  >
    Nothing more to show
  </div>
</div>
`;
